doc/man/aboot.8
.\" This manpage has been automatically generated by docbook2man-spec
.\" from a DocBook document.  docbook2man-spec can be found at:
.\" <http://shell.ipoline.com/~elmert/hacks/docbook2X/> 
.\" Please send any bug reports, improvements, comments, patches, 
.\" etc. to Steve Cheng <steve@ggi-project.org>.
.TH "ABOOT" "8" "11 Juli 2002" "aboot" ""
.SH NAME
aboot \- The secondary boot loader for Linux/Alpha
.SH "COPYRIGHT"
.PP
\fBaboot\fR is Copyright (C) 1996 Linus Torvalds, David Mosberger-Tang and Michael Schwingen
.SH "DESCRIPTION"
.PP

\fBaboot\fR is the secondary boot loader needed if
the Linux/Alpha system is booted via SRM. It is usually installed during
system installation by \fBswriteboot\fR(8). 
.PP
To automate the booting process predefined boot configurations can be stored
in \fIetc/aboot.conf\fR(5). 
.PP
\fBaboot\fR can also be used interactively by issuing
.PP
\fBboot \fIdka0 -fl "i"\fB\fR
.PP
at the SRM-Prompt where dka0 has to be replaced by the device used. 
.PP
In the interactive mode you can use the command h to show your current
\fIetc/aboot.conf\fR and then the appropriate number
(or a complete boot command as in SRM) to boot Linux.
.PP
If you do not want to use the \fIetc/aboot.conf\fR as set
up by \fBabootconf\fR you can explicitly tell 
\fBaboot\fR where to look for \fIetc/aboot.conf\fR by prepending the number of the configuration with the proper
partition, i.e. if your \fIetc/aboot.conf\fR is on your
second partition and you want to boot the entry labeled three you would
issue
.PP
\fBboot \fIdqa -fl "2:3"\fB\fR
.PP
Note that this applies to Linux booting via SRM only as
VMS, True64, *BSD and
Windows NT each have their own secondary boot loader.
Furthermore \fBaboot\fR is also not needed if the system is booted via
\fBmilo\fR.
.SH "AUTHOR"
.PP
This man page was written by Helge Kreutzmann
<kreutzm@itp.uni-hannover.de> for the Debian GNU/Linux project
but may be used by others.
.SH "FILES"
.PP
\fIetc/aboot.conf\fR
.SH "SEE ALSO"
.PP
\fBabootconf\fR(8), \fIaboot.conf\fR(5),
\fBswriteboot\fR(8), HP SRM Manual


ChangeLog
2003-03-21  Will Woods <will.woods@hp.com>
	* Added include/string.h, since we have our own string.c, replaced 
	  #includes of <linux/string.h> with "string.h" since linux/string.h
	  no longer has user-includable prototypes.
	  (thanks to Kelledin <kelledin@skarpsey.dyndns.org> for the idea)
	* Changed #includes of our headers to use quotes instead of <>
	* Don't build with -mcpu=ev4 (default code works on ev4 and is smaller)
	* Cleaned up a couple other warnings 

2003-01-14  Will Woods <will.woods@hp.com>
	* Fixed typos in man pages ("True 64")
	* Bumped version number to 0.9b
	* Released 0.9b. Hooray!

2002-09-18  Will Woods <will.woods@hp.com>
	* branched the 0.9 tree to add bugfixes
	* fixed some compilation errors with newer gcc/glibc versions
	* added newer manpages

2001-11-14  Will Woods <will.woods@compaq.com>
        * sdisklabel/sdisklabel.c: fixed my fix to the cylinder counting bug.
          oops. also added (currently undocumented) "force" flag, to skip
          error checking in sdisklabel.

2001-10-26  Will Woods <will.woods@compaq.com>
        * Released aboot 0.9. Yay!

2001-10-25  Will Woods <will.woods@compaq.com>
	* sdisklabel/sdisklabel.c: fixed cylinder-counting bug

2001-10-11  Will Woods <will.woods@compaq.com>
	* applied a patch from the debian version of aboot. Here's the debian
	maintainer's changelog:

  * New upstream (from CVS)
  * I am calling this version of aboot from CVS 0.8.
  * Minor tweeks to make it compile with kernel 2.4.x headers.
  * e2writeboot now returns non-zero on error.  Closes: #107757
  -- Gregory W. Johnson <gjohnson@physics.clarku.edu>  Wed, 15 Aug 2001

	Since he called his version 0.8, I will call this the 0.9 tree, to 
	avoid confusion. Hence this is aboot 0.9pre1. 

2001-10-09  Will Woods <will.woods@compaq.com>
	* aboot.c, disk.c, include/config.c: replaced ABOOT_MINOR and 
	ABOOT_MAJOR with ABOOT_VERSION. Bumped version number to 0.8pre2.
	* aboot.c: fixed ELF header bug that snuck in somewhere.
	* fs/ext2.c, Makefile: fixes for proper #includes to allow compilation
	of userspace testing aboot against the 2.4 kernel.

2001-10-08  Will Woods <will.woods@compaq.com>
	* fs/ext2.c: Fixed elusive overflow bug. If part of the kernel was 
	past the 524,287th block (given 4096 bytes per block) the byte offset
	was too big to fit in an int. 
	* sdisklabel/sdisklabel.c: borrowed code from fdisk for getting
	the correct size of the disk. 
	
2001-07-10  Will Woods <will.woods@compaq.com>
	* aboot.c, tools/objstrip.c: elf_check_arch takes different arguments
	in Linux 2.4 than it did in 2.2; added an aboot_elf_check_arch which
	does the right thing based on version info from linux/version.h. 
	* disk.c: a function called get_options exists in 2.4 kernels, so
	aboot's get_options function was renamed get_aboot_options(). 
	(both changes based on a patch by Christian Groessler <cpg@aladdin.de>)

2001-04-26  Will Woods <will.woods@compaq.com>
	* disk.c: changed /etc/aboot.conf and aboot.conf back to their
	original settings. If it ain't broke, why fix it?
	* added the man pages in doc/man, written by Helge Kreutzmann
	<kreutzm@itp.uni-hannover.de> (Thanks Helge!)

2001-03-20  Will Woods <will.woods@compaq.com>
	* disk.c: swapped position of /etc/aboot.conf and aboot.conf in the 
	configs array, making /aboot.conf the default.
	* isolib.c: added parse_rock_ridge_inode and get_rock_ridge_symlink
	adapted from rock.c in the linux kernel source. 
	* isolib.c: added iso_follow_link and iso_readdir, adapted from
	fs/ext2.c
        * made minor changes here and there to quiet gcc warnings - 
	initializing variables to 0, casting variables properly, etc

2000-06-27  David Huggins-Daines  <dhd@linuxcare.com>

	* sdisklabel/sdisklabel.c (main): Off-by-one error in argument
	checking (not my fault!).
	* disk.c (load_raw): This didn't work at all.  I've changed it to
	use an uncompressed kernel with headers.  Eventually it will be
	fixed so that it can accept compressed kernels as well (by
	checking for gzip magic numbers).
	* lib/vsprintf.c (vsprintf): Patch from Matt Wilson, fix the 'h'
	format in vsprintf().
	* utils.c: Patch from Matt Wilson <msw@redhat.com>, to allow
	compiling against linux 2.3 header files.

2000-05-02  David Huggins-Daines  <dhd@linuxcare.com>

	* Makefile: Added the ability to build a debugging aboot that runs
	as a userspace program.  Great for filesystem code debugging.
	Also, optimize for size instead of speed.
	* aboot.c: Remove a.out support, add multiple ELF pheader
	support.  Also make the parameter offset, and kernel stack offset
	non-hardcoded in order to deal with new, bigger consoles.
	* disk.c: Change "plain" to "uncompressed" and load_aout() to
	load_uncompressed().  Implement multiple pheaders for raw boot
	(untested) and uncompressed boot (tested). Also look for
	alternative config files (needed for CD booting and such)
	* fs/ext2.c: Cache the root inode, and do proper 64-bit arithmetic
	in ext2_iget() so that >2G filesystems work reliably (doh!)
	* head.S: make run_kernel() assembly in order to save a few bytes.
	* include/aboot.h: prototype run_kernel as attribute((noreturn)),
	maybe this will save a few bytes too.
	* fs/ufs.c: Implement dummy fstat().  I don't really know if the
	UFS code still works, and I'm tempted to remove it.
	* include/cons.h: Add a definition of the CTB, though we don't
	really use it yet.
	* include/iso.h: Rock Ridge constants
	* lib/isolib.c: Rock Ridge filenames, code cleanup, and fix the
	bug where 2048-byte blocks don't work.
	* zip/misc.c: necessary stuff for multiple pheaders.

2000-04-10  David Huggins-Daines  <dhuggins@linuxcare.com>

	* lib/isolib.c (iso_iget): Gar!  More stupidity in the ISO code,
	now fixed (see comment about *pnt being overwritten).  With luck
	this should be 0.7 final, so I won't need to do another 0.6

2000-04-02  David Huggins-Daines  <dhuggins@linuxcare.com>

	* lib/isolib.c: Fix some obvious stupidity in the ISO code (could

aboot.c
/*
 * aboot.c
 *
 * This file is part of aboot, the SRM bootloader for Linux/Alpha
 * Copyright (C) 1996 Linus Torvalds, David Mosberger, and Michael Schwingen.
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License as
 * published by the Free Software Foundation; either version 2 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 */

#include <asm/console.h>
#include "hwrpb.h"
#include "system.h"

#include <elf.h>
#include <errno.h>

#include "aboot.h"
#include "config.h"
#include "cons.h"
#include "setjmp.h"
#include "utils.h"
#include <string.h>

struct bootfs *	bfs = 0;		/* filesystem to boot from */
char *		dest_addr = 0;
jmp_buf		jump_buffer;

unsigned long	start_addr = START_ADDR; /* default only */
struct segment* chunks;
int             nchunks;

char		boot_file[256] = "";
char		initrd_file[256] = "";
char		kernel_args[256] = "";
char *		bss_start = 0;
long		bss_size = 0;		 /* size of bss */
unsigned long   initrd_start = 0;
unsigned long   initrd_size = 0;

/* Offsets from start_addr (yes, they are negative, because start_addr
   is actually the load address plus the text offset) */
#define PARAM_OFFSET	-0x6000 /* load address + 0xa000 */
#define STACK_OFFSET	-0xe000 /* load address + 0x2000 */

/* eventually we may have to deal with 44-bit physical addressing and
   also possibly larger pages */
unsigned long	page_offset = 0xfffffc0000000000;
unsigned long   page_shift  = 13;

static unsigned long entry_addr = START_ADDR;

/*
 * The decompression code calls this function after decompressing the
 * first block of the object file.  The first block must contain all
 * the relevant header information.
 */
long
first_block (const unsigned char *buf, long blocksize)
{
	Elf64_Ehdr *elf;
	Elf64_Phdr *phdrs;
	int i, j;

	elf  = (Elf64_Ehdr *) buf;
	
	if (elf->e_ident[0] != 0x7f
	    || elf->e_ident[1] != 'E'
	    || elf->e_ident[2] != 'L'
	    || elf->e_ident[3] != 'F')
	{
		/* Fail silently, it might be a compressed file */
		return -1;
	}
	if (elf->e_ident[EI_CLASS] != ELFCLASS64
	    || elf->e_ident[EI_DATA] != ELFDATA2LSB
	    || elf->e_machine != EM_ALPHA)
	{
		printf("aboot: ELF executable not for this machine\n");
		return -1;
	}

	/* Looks like an ELF binary. */
	if (elf->e_type != ET_EXEC) {
		printf("aboot: not an executable ELF file\n");
		return -1;
	}

	if (elf->e_phoff + elf->e_phnum * sizeof(*phdrs) > (unsigned) blocksize)
	{
		printf("aboot: "
		       "ELF program headers not in first block (%ld)\n",
		       (long) elf->e_phoff);
		return -1;
	}

	phdrs = (Elf64_Phdr *) (buf + elf->e_phoff);
	chunks = malloc(sizeof(struct segment) * elf->e_phnum);
	start_addr = phdrs[0].p_vaddr; /* assume they are sorted */
	entry_addr = elf->e_entry;

#ifdef DEBUG
	printf("aboot: %d program headers, start address %#lx, entry %#lx\n",
	       elf->e_phnum, start_addr, entry_addr);
#endif

	for (i = j = 0; i < elf->e_phnum; ++i) {
		int status;

		if (phdrs[i].p_type != PT_LOAD)
			continue;

		chunks[j].addr   = phdrs[i].p_vaddr;
		chunks[j].offset = phdrs[i].p_offset;
		chunks[j].size   = phdrs[i].p_filesz;

#ifdef DEBUG
		printf("aboot: PHDR %d vaddr %#lx offset %#lx size %#lx\n",
		       i, chunks[j].addr, chunks[j].offset, chunks[j].size);
#endif

#ifndef TESTING
		status = check_memory(chunks[j].addr, chunks[j].size);
		if (status) {
			printf("aboot: Can't load kernel.\n"
			       "  Memory at %lx - %lx (PHDR %i) "
			         "is %s\n",
			       chunks[j].addr,
			       chunks[j].addr + chunks[j].size - 1,
			       i,
			       (status == -ENOMEM) ?
			          "Not Found" :
				  "Busy (Reserved)");
			return -1;
		}
#endif

		if (phdrs[i].p_memsz > phdrs[i].p_filesz) {
			if (bss_size > 0) {
				printf("aboot: Can't load kernel.\n"
				       "  Multiple BSS segments"
				       " (PHDR %d)\n", i);
				return -1;
			}
			bss_start = (char *) (phdrs[i].p_vaddr +
				              phdrs[i].p_filesz);
			bss_size = (phdrs[i].p_memsz - phdrs[i].p_filesz);
		}

		j++;
	}
	nchunks = j;
#ifdef DEBUG
	printf("aboot: bss at 0x%p, size %#lx\n", bss_start, bss_size);
#endif

	return 0;
}

static void
get_boot_args(void)
{
	long result;

	/* get boot command line: */
#ifdef TESTING
	const char *e;
	if ((e = getenv("BOOTED_FILE"))) {
		strncpy(boot_file, e, sizeof(boot_file)-1);
		boot_file[sizeof(boot_file)-1] = 0;
	} else {
		strcpy(boot_file, "vmlinux.gz");
	}
	if ((e = getenv("BOOTED_OSFLAGS"))) {
		strncpy(kernel_args, e, sizeof(kernel_args)-1);
		kernel_args[sizeof(kernel_args)-1] = 0;
	} else {
		strcpy(kernel_args, "i");
	}
#else
	result = cons_getenv(ENV_BOOTED_FILE, boot_file, sizeof(boot_file));
	if (result < 0) {
		printf("aboot: warning: can't get ENV_BOOTED_FILE "
		       "(result=%lx)!\n", result);
		strcpy(boot_file, "vmlinux.gz");
	}
	result = cons_getenv(ENV_BOOTED_OSFLAGS,
			     kernel_args, sizeof(kernel_args));
	if  (result < 0) {
		printf("aboot: warning: can't get ENV_BOOTED_OSFLAGS "
		       "(result=%lx)!\n", result);
		strcpy(kernel_args, "i");
	}
#endif /* TESTING */
}

#ifdef TESTING
long config_file_partition = 1;
void halt()
{
	exit(0);
}

void unzip_error(char *x)
{
	printf("unzip: %s\n", x);
}


int main()
{
	extern long load_kernel();
	long result;

	get_boot_args();
	result = load_kernel();
	if (result < 0) {
		printf("aboot: kernel load failed (%ld)\n", result);
		return 0;
	}
	printf("aboot: starting kernel %s with arguments %s\n",
	       boot_file, kernel_args);
	return 0;
}
#else /* not TESTING */
/*
 * Head transfers control to this function.  Don't call it main() to avoid
 * gcc doing magic initialization things that we don't want.
 */
void
main_ (void)
{
	extern long load_kernel (void);
	long i, result;

	cons_init();

	printf("aboot: Linux/Alpha SRM bootloader version "ABOOT_VERSION"\n");

	/* don't know how to deal with this yet */
	if (INIT_HWRPB->pagesize != 8192) {
		printf("aboot: expected 8kB pages, got %ldkB\n",
		       INIT_HWRPB->pagesize >> 10);

		cons_close_console();
		return;
	}

	pal_init();
	get_boot_args();
	result = load_kernel();
	if (result < 0) {
		printf("aboot: kernel load failed (%ld)\n", result);
		cons_close_console();
		return;
	}
	printf("aboot: starting kernel %s with arguments %s\n",
	       boot_file, kernel_args);
	strcpy((char*)start_addr + PARAM_OFFSET, kernel_args);
	*(unsigned long *)(start_addr + PARAM_OFFSET + 0x100)
		= initrd_start;
	*(unsigned long *)(start_addr + PARAM_OFFSET + 0x108)
		= initrd_size;

	cons_close_console();
	run_kernel(entry_addr, start_addr + STACK_OFFSET);

	cons_open_console();
	printf("aboot: kernel returned unexpectedly.  Halting slowly...\n");
	for (i = 0 ; i < 0x100000000 ; i++)
		/* nothing */;
	cons_close_console();
	halt();
}
#endif /* TESTING */
